### DISCUSSION TITLE
ChooserViewSet and InlinePanel

### DISCUSSION THREAD

**Discussion by kiyoshi-satoo** [wagtail/wagtail#12029] [Q&A]
models.py

```python
class HeroWhyItem(Orderable):
    hero_why = ParentalKey('core.HeroWhy', on_delete=models.CASCADE, related_name='hero_why_items')
    title = models.CharField(max_length=120, help_text=_('Title of the item'), verbose_name=_('Title'))
    description = models.TextField(help_text=_('Description of the item'), verbose_name=_('Description'))

    panels = [
        FieldPanel('title'),
        FieldPanel('description'),
    ]

    def __str__(self):
        return f'{self.title}'

class HeroWhy(index.Indexed, ClusterableModel):
    name = models.CharField(max_length=120, help_text=_('Use this for variant name'), verbose_name=_('Name'))
    panels = [
        FieldPanel("name"),
        InlinePanel("hero_why_items", label=_("Card"), heading=_('Slides'), min_num=1, max_num=4)
    ]

    search_fields = [
        index.SearchField('name'),
        index.AutocompleteField('name'),
    ]

    class Meta:
        verbose_name = _("Hero Why")
        verbose_name_plural = _("Hero Whys")

    def __str__(self):
        return f'{self.name}'
```

views.py

```python
class HeroWhyChooserViewSet(ChooserViewSet):
    model = "core.HeroWhy"
    choose_one_text = "Choose"
    choose_another_text = "Choose another"
    edit_item_text = "Edit this"
    form_fields = ["name", "hero_why_items"]

hero_why_chooser_viewset = HeroWhyChooserViewSet('hero_why_chooser')
```

wagtail_hooks.py

```python
@hooks.register('register_admin_viewset')
def register_hero_why_chooser_viewset():
    return hero_why_chooser_viewset

```

I get an error when opening a chooser modal
`django.core.exceptions.FieldError: Unknown field(s) (hero_why_items) specified for HeroWhy`

The question is the following. How correct is this decision ? And is it possible to customise ChooserViewSet to work with Inline fields ?

Any help is appreciated!

**Comment by kiyoshi-satoo** [comment]
In addition.ChooserViewSet does not work with images (No widget for image selection) and Inline models

**Comment by Ganesh-403** [comment]
@kiyoshi-satoo The error:

```text
django.core.exceptions.FieldError:
Unknown field(s) (hero_why_items) specified for HeroWhy
```

is expected because `hero_why_items` is **not a Django model field** on `HeroWhy`. It's the **reverse relation** created by the `ParentalKey`, which `InlinePanel` knows how to work with, but a Django `ModelForm` (which `ChooserViewSet` uses for its creation form) does not. `form_fields` can only contain actual model fields. :contentReference[oaicite:0]{index=0}

So in your example:

```python
form_fields = ["name", "hero_why_items"]
```

`name` is valid, but `hero_why_items` is not.

Regarding your broader question:

> Is it possible to customise `ChooserViewSet` to work with `InlinePanel`?

From what I could find, there isn't built-in support for this. The "Create" tab of `ChooserViewSet` is backed by a standard model form (or a custom `creation_form_class`), whereas `InlinePanel` relies on Wagtail's edit handler machinery and formsets (`ClusterForm`) to manage child objects. The documented customization points for `ChooserViewSet` are `creation_form_class`, `form_fields`, and `exclude_form_fields`; I couldn't find any support for automatically rendering `InlinePanel`s or nested chooser widgets in the modal. :contentReference[oaicite:1]{index=1}

The same applies to your follow-up observation about image fields. A plain Django form won't automatically render Wagtail chooser widgets unless the appropriate form/edit-handler integration is used.

If your goal is simply to create a `HeroWhy` from the chooser, I'd recommend limiting the creation form to the parent fields, for example:

```python
form_fields = ["name"]
```

and then editing the `HeroWhy` object afterwards (or via a `ModelViewSet`/snippet edit view) where the `InlinePanel` is available.

If your requirement is to create the parent object **and** its `HeroWhyItem` children in a single chooser modal, I suspect that would require a custom implementation (for example, a custom `creation_form_class` together with a custom create view that manages formsets). I couldn't find an existing extension point that provides this behavior out of the box.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.
